lib/dat-worker-pool/runner.rb

```diff
@@ -71,6 +66,30 @@ def build_worker
       @worker_class.new(self, @queue).tap(&:dwp_start)
     end
 
+    # the workers should be told to shutdown before the queue because the queue
+    # shutdown will wake them up; a worker popping on a shutdown queue will
+    # always get `nil` back and will loop as fast as allowed until its shutdown
+    # flag is flipped, so shutting down the workers then the queue keeps them
+    # from looping as fast as possible; use an until loop instead of each to
+    # join all the workers, while we are joining a worker a different worker can
+    # shutdown and remove itself from the `@workers` array; rescue when joining
+    # the workers, ruby will raise any exceptions that aren't handled by a
+    # thread when its joined, this allows all the workers to be joined
+    def graceful_shutdown(timeout)
+      @workers.each(&:dwp_shutdown)
+      @queue.shutdown
+      OptionalTimeout.new(timeout) do
+        until @workers.empty?
+          worker = @workers.first
+          begin
+            worker.dwp_join
+          rescue StandardError
+            self.remove_worker(worker)
```

Comment:
@jcredding also how do you handle every worker having an exception on join?  It looks like here you ignore worker join exceptions and just remove them and go on.  This means that the only standard errors that instigate a forced shutdown are standard errors during calling `dwp_shutdown` on the workers and calling `shutdown` on the queue.  Maybe I am missing something though and this is intended?